hfile/reader.go

```diff
@@ -254,3 +265,74 @@ func (r *Reader) GetIterator() *Iterator {
 		return NewIterator(r)
 	}
 }
+
+// TODO(benjy): Write tests for this, once the writer supports writing the FileInfo block.
+func (r *Reader) readFileInfo(data []byte) (err error) {
+	type outOfBounds struct {}
+
+	defer func() {
+		if r := recover(); r != nil {
+			switch r.(type) {
+			case outOfBounds:
+				err = errors.New("Tried to read past end of FileInfo block!")
+			default:
+				panic(r)
+			}
+		}
+	}()
+
+	r.FileInfo.InfoFields = make(map[string]string)
+
+	fileInfoBlockEnd := r.DataIndexOffset
+
+	p := r.FileInfoOffset
+
+	if p == fileInfoBlockEnd {
+		log.Println("[Reader.readFileInfo] No FileInfo block found. Skipping.")
+		return nil
+	}
+
+	// Safely move the read pointer.
+	advance := func(increment int) {
```

Comment:
Style nit: I don't generally see many local function values unless they are used as values (eg passing them to something). I'd pull it out, if you need it, and then you can test it on its own too if you wanted to.

I'm also curious if we could just use the stdlib's `bytes.Reader` as we do elsewhere when we want a stateful pointer into a []byte with bounds checking (eg in the `Trailer` reading above).